fl4health/utils/data_generation.py

```diff
@@ -0,0 +1,295 @@
+        components of W and b. Note that sampling occurs for EACH client independently. The larger alpha and beta
+        the larger the variance in these values, implying higher probability of heterogeneity.
+
+        Returns:
+            List[Tuple[torch.Tensor, torch.Tensor]]: Set of input and output tensors for each client.
+        """
+        tensors_per_client: List[Tuple[torch.Tensor, torch.Tensor]] = []
+        for _ in range(self.num_clients):
+            B = torch.normal(0.0, self.beta, (1,))
+            # v_k in the FedProx paper
+            input_means = torch.normal(B, torch.ones(self.input_dim))
+
+            # u_k in the FedProx paper
+            affine_transform_means = torch.normal(0, self.alpha, (1,)).item()
+            client_X, client_Y = self.get_input_output_tensors(
+                affine_transform_means, input_means, self.input_covariance
+            )
+            tensors_per_client.append((client_X, client_Y))
+        return tensors_per_client
+
+
+class SyntheticIidFedProxDataset(SyntheticFedProxDataset):
+    def __init__(
+        self,
+        num_clients: int,
+        temperature: float = 1.0,
+        input_dim: int = 60,
+        output_dim: int = 10,
+        samples_per_client: int = 1000,
+    ) -> None:
+        """
+        IID Synthetic dataset generator modeled after the implementation in the original FedProx paper. See Appendix
+        C.1 in the paper link below for additional details. The IID generation code is based strictly on the
+        description in the appendix for IID dataset generation.
+
+        Paper link: https://arxiv.org/abs/1812.06127
+
+        NOTE: This generator ends up with fairly skewed labels in generation. That is, many of the clients will not
+        have representations of all the labels. This has been verified as also occurring in the reference code above
+        and is not a bug.
+
+        Args:
+            num_clients (int): Number of datasets (one per client) to generate
+            temperature (float, optional): temperature used for the softmax mapping to labels. Defaults to 1.0.
+            input_dim (int, optional): dimension of the input features for the synthetic dataset. Default is as in the
+                FedProx paper. Defaults to 60.
+            output_dim (int, optional): dimension of the output labels for the synthetic dataset. These are one-hot
+                encoding labels. Default is as in the FedProx paper. Defaults to 10.
+            samples_per_client (int, optional): Number of samples to generate in each client's dataset.
+                Defaults to 1000.
+        """
+        super().__init__(
+            num_clients=num_clients,
+            temperature=temperature,
+            input_dim=input_dim,
+            output_dim=output_dim,
+            samples_per_client=samples_per_client,
+        )
+
+        # As described in the original paper, the affine transformation is SHARED by all clients and the elements
```

Comment:
Just to highlight that `v` is also the same across clients, maybe, for clarity, we can put this here `self.v = torch.zeros(self.input_dim)`, and in the `get_input_output_tensors` function use it `loc = self.v`

Reply:
I went a little bit further and just instantiated the multivariate normal in the init function to emphasis that it is also shared. Let me know what if that makes sense.

Reply:
Looks great!